File: payload_ltx.js
/* LTX  Payload Decoder (Uplink)
* V1.17 (C) JoEmbedded.de 
* https://github.com/joembedded/payload-decoder
* LTX uses a flexible, compressed format, see LTX-Documentation
* using FLoat32 and Float16 on fPort 1-199 (fPort sets the 'units')
*/
//REMOVE-START
/* 
Installation (ChirpStack and TTN V3:
* - Remove block 'TEST-CONSOLE'
* 
* Test: ...\payloaddecoder> ./qjs payload_ltx.js
*/
//REMOVE-END

// ChirpStack V4
function decodeUplink(input) { // assume input is valid
    if ((input.fPort > 0) && (input.fPort < 200)) {
        // fPort 1-199: LTX Standard-Uplink, Port defines 'known types'
        return { data: Decoder(input.bytes, input.fPort) }
    } else {
        return { data: { errors: [`LTX: fPort:${input.fPort} unknown`] } };
    }
}

// TTN / Chirpstack
function Decoder(bytes, port) {
    const decoded = {};
    // Byte 0: Hello-Flags for FLAGS and REASON
    let ianz = bytes.length;
    if (ianz < 1) return { errors: ["LTX: Payload len < 1"] };
    if (port < 1 || port > 199) return { errors: [`LTX: fPort:${port} unknown`] };

    // Transfer input bytes to BinaryArray
    const view = new DataView(new ArrayBuffer(ianz));
    let cursor = 0;
    bytes.forEach((b, i) => view.setUint8(i, b));
    const flags = view.getUint8(cursor++);
    ianz--;
    decoded.flags = "";
    if (flags & 128) decoded.flags += "(Reset)";
    if (flags & 64) decoded.flags += "(Alarm)";
    else if (flags & 32) decoded.flags += "(old Alarm)";
    if (flags & 16) decoded.flags += "(Measure)";

    const reason = flags & 15;
    if (reason === 2) decoded.reason = "(Auto)";
    else if (reason === 3) decoded.reason = "(Manual)";
    else decoded.reason = `(Reason:${reason}?)`;

    // Channels:0..89, HK:90..99(max 127) - Decoder V1.x does not care extended itoks >= 128!
    let ichan = 0;
    const typeunits = deftypes[port];
    let dtidx = 0;  // Index in deftypes
    decoded.chans = [];
    while (ianz-- > 0 && ichan < 128) {
        const itok = view.getUint8(cursor++);
        if (!(itok & 128)) {
            let cmanz = itok & 63; // Channels: number of consecutive values
            if (!cmanz) {   // cmanz=0: New channel following
                if (ianz < 1) return { errors: ["LTX: Format(1) invalid"] };
                ichan = view.getUint8(cursor++); // AutoEnd for ichan >= 128
                ianz--;
                break;
            }
            const f16 = itok & 64;
            while (cmanz-- > 0) {
                const puob = { channel: ichan };
                if (typeunits && typeunits.length && typeunits.length > 0) { // Known Type with units
                    puob.unit = typeunits[dtidx % (typeunits.length)];
                }
                if (!f16) { // Following cmanz Float32
                    if (ianz < 4) return { errors: ["LTX: Format(2) invalid"] };
                    puob.prec = "F32";
                    const u32 = view.getUint32(cursor, false); // Get Float32-Bits
                    if ((u32 >>> 23) == 0x1FF) puob.msg = getLTXError(u32 & 0x7FFFFF);
                    else puob.value = parseFloat(view.getFloat32(cursor, false).toPrecision(8));
                    cursor += 4;
                    ianz -= 4;
                } else { // Following cmanz Float16
                    if (ianz < 2) return { errors: ["LTX: Format(3) invalid"] };
                    puob.prec = "F16";
                    const u16 = view.getUint16(cursor, false); // Get Float16-Bits
                    if ((u16 >>> 10) == 0x3F) puob.msg = getLTXError(u16 & 1023);
                    else  puob.value = parseFloat(view.getFloat16(cursor, false));
                    cursor += 2;
                    ianz -= 2;
                }
                if (('value' in puob) && !Number.isFinite(puob.value)) {
                    delete puob.value;
                    puob.msg = getLTXError(0);
                }
                decoded.chans.push(puob);
                dtidx++;
                ichan++;
            }
        } else { // HK, channels 90.. 
            if (ichan < 90) ichan = 90;
            let cbits = itok & 127; // HK: 7-bit mask of consecutive values
            for(let i=0;i<7;i++){
                if (cbits & 1) {
                    if (ianz < 2) return { errors: ["LTX: Format(4) invalid Data"] };
                    const puob = { channel: ichan, prec: "F16", unit: unitDescrHK(ichan) };
                    const u16 = view.getUint16(cursor, false); // Get Float16-Bits
                    if ((u16 >>> 10) == 0x3F) puob.msg = getLTXError(u16 & 1023);
                    else {
                        puob.value = parseFloat(view.getFloat16(cursor, false));
                        if(!Number.isFinite(puob.value)) {
                            delete puob.value;
                            puob.msg = getLTXError(0);
                        }
                    }
                    decoded.chans.push(puob);
                    cursor += 2;
                    ianz -= 2;
                }
                ichan++;
                cbits >>>= 1;
            }
        }
    }
    return decoded;
}
// List of 'known types' with (opt. repeated) units - LTX-Sensors
const deftypes = {
    // 1-9 free or for  custom sensors
    10: ['°C'], // 10: All channels are Temperatures
    11: ['%rH', '°C'], // 11: rH/T-Sensor
    12: ['Bar', '°C'], // 12: Pressure/Level Sensor Bar
    13: ['m', '°C'], // 13: Level Sensor m
    14: ['m', 'dBm'], // 14: Distance(s) Sensor (Radar)
    15: ['°C', 'uS/cm'], // 15: Water Conductivity (ES-2)
};

// descHK - Unit and Description of HK-Channels
function unitDescrHK(ichan) {
    switch (ichan) {
        case 90: return "V(HK_Bat)";
        case 91: return "°C(HK_intTemp)";
        case 92: return "%rH(HK_intHum.)";
        case 93: return "mAh(HK_usedEnergy)";
        case 94: return "mBar(HK_Baro)";
        default: return "HK_unknown";
    }
}
// LTX Standard Error Codes, included in F16 or F32
function getLTXError(errno) {
    switch (errno) {
        case 0: return "NumberOverflow"; // CodecError
        case 1: return "NoValue";
        case 2: return "NoReply";
        case 3: return "OldValue";
        // 4,5 unused
        case 6: return "ErrorCRC";
        case 7: return "DataError";
        case 8: return "NoCachedValue";
        default: return `Err${errno}`;
    }
}
// Only a few JS engines support Float16
if (!DataView.prototype.getFloat16) {
    DataView.prototype.getFloat16 = function (byteOffset, littleEndian) {
        const uint16 = this.getUint16(byteOffset, littleEndian);
        const sign = (uint16 & 0x8000) ? -1 : 1;
        const exponent = (uint16 >>> 10) & 0x1F;
        const fraction = uint16 & 0x03FF;
        if (exponent === 0) return sign * Math.pow(2, -14) * (fraction / Math.pow(2, 10)); //Subnormal
        else if (exponent === 0x1F) return fraction ? NaN : sign * Infinity; // NaN or Infinity
        return (sign * Math.pow(2, exponent - 15) * (1 + fraction / Math.pow(2, 10))).toPrecision(6);
    }
}

//REMOVE-START ----------------- TEST-CONSOLE, for TTN / ChirpStack remove the following parts: -----------------
// helper to pass test data to/from decoder
function testDecoder(hexString, port) {
    const msg = hexString.replace(/\s/g, '');
    testBytes = [];
    for (let i = 0; i < msg.length; i += 2) {
        testBytes.push(parseInt(msg.substring(i, i + 2), 16));
    }
    const theObject = decodeUplink({ fPort: port, bytes: testBytes });
    return JSON.stringify(theObject, null, 2);
}

function main() {
    console.log("---TEST-DECODER---");

    // Test-Payloads (in HEX), may contain spaces for better readability
    // taken from LTX-INTENT Data Logger (Type 1500) with 1 or 3 SDI-12-Sensors
    const testmsg = [

        // "13 01 41A4E148  9F 42A3 4DA8 5172 2B3F 63E8", // Manual, Value: 20.610001, HKs: 3.31836(V), 22.6250(°C), 43.5625(%), 0.0566101(mAh), 1012.00(mBar)
        // "12 42 4D24 4CE4 01 41948F5C 88 355B", // Auto 2 Values(F16): 20.5625, 19.5625, Value(F32): 18.570000 , HK: 0.334717(mAh)
        // "7242FC02FC0201FF800002884479", // Alarm, Auto, 3 Values: 'No Reply', HK: 4.47266(mAH)
        // "93585B2F5AE25A965A3A59ED59A0595058FC58B558685813577F56EC5649559F54F45460538C5243511E4F6B4D2E4884BAE4", // Manual,Reset, 24*F16-Values
        "12585bb0000000003c005c7b72b67c007c005bb05599564456e25784580c586258b558fc594f599c59ee5a3c5a8c5ae45b29" // With INF.16 in Data om Chan 6/7
    ]; 
    const testport = 1

    testmsg.forEach(e => {
        console.log("----Test-Payload:-----");
        console.log(e, " Port:", testport);
        console.log("----JSON Result:-----");
        console.log(testDecoder(e, testport));
    })
}

main();
//REMOVE-END ---

File: payload_ltx.exs
# LTX Payload Decoder (Uplink) in Elixir auf ELEMENT IoT
# V1.16 (C) JoEmbedded.de
# https://github.com/joembedded/payload-decoder

# Aktionen vor Speichern fuer ELEMENT IoT
# - 'use Platform.Parsing.Behaviour' einkommentieren
# - Block zwischen ...TEST_START und ...TEST_END entfetneen
# - Letzte Zeile 'Parser.main()' auskommentieren
#
# ELEMENT IoT verweigert das Speichern des Parsers ohne irgendwelche Angaben,
# sofern IO. etc. vorhanden ist. Auch verweidgert ELEMENT viele
# Sprachkonstrikte, z.B. divese map-Operationen... Die Liste ist lang...

defmodule Parser do
  # use Platform.Parsing.Behaviour
  import Bitwise

  # ---PARSE---
  def parse(<<flags, rest::binary>>, meta) do
    %{meta: %{frame_port: fPort}} = meta

    # Grund für die Übertragung
    reason =
      (flags &&& 0x0F)
      |> case do
        2 -> "(Auto)"
        3 -> "(Manual)"
        r -> "(Reason:#{r}?)"
      end

    # Flags extrahieren
    flag_str =
      [
        if((flags &&& 0x80) > 0, do: "(Reset)"),
        if((flags &&& 0x40) > 0, do: "(Alarm)"),
        if((flags &&& 0x20) > 0, do: "(old Alarm)"),
        if((flags &&& 0x10) > 0, do: "(Measure)")
      ]
      |> Enum.filter(& &1)
      |> Enum.join()

    # Rest ist Channel Payload
    decpay = decode_channels(rest, fPort)

    # Ergebnis zurückgeben, port nur informativ
    %{
      port: fPort,
      reason: reason,
      flags: flag_str,
      chans: decpay
    }
  end

  # Channel Payload
  def decode_channels(pbytes, port) do
    decode_channels(pbytes, 0, [], port)
  end

  def decode_channels(<<>>, _ichan, result, _port) do
    result
  end

  # Kanal folgt
  def decode_channels(<<itok, rest::binary>>, _ichan, result, port) when itok == 0 do
    # Neuer Kanal folgt
    <<new_ichan, rest::binary>> = rest
    decode_channels(rest, new_ichan, result, port)
  end

  # 1-63 F32 folgen
  def decode_channels(<<itok, rest::binary>>, ichan, result, port) when itok < 64 do
    # Float32-Daten
    anzf32 = itok
    decode_f32(anzf32, rest, ichan, result, port)
  end

  # 0-63 F16 folgen
  def decode_channels(<<itok, rest::binary>>, ichan, result, port) when itok < 128 do
    # Float16-Daten
    anzf16 = itok &&& 63
    decode_f16(anzf16, rest, ichan, result, port)
  end

  # 0-7 HK folgen
  def decode_channels(<<itok, rest::binary>>, ichan, result, port) do
    # HK als Float16-Daten
    hk16bits = itok
    ichan = if ichan < 90, do: 90, else: ichan
    decode_hk(hk16bits, 7, rest, ichan, result, port)
  end

  # n * F32 entnehmen
  def decode_f32(0, rest, ichan, result, port) do
    decode_channels(rest, ichan, result, port)
  end

  def decode_f32(anzf32, <<0x1FF::size(9), errno::size(23), rest::binary>>, ichan, result, port)
      when anzf32 > 0 do
    errmsg = ltx_error(errno)
    unit = ltx_units(port, length(result))
    cv = %{channel: ichan, msg: errmsg, unit: unit, prec: "F32"}

    result = result ++ [cv]
    decode_f32(anzf32 - 1, rest, ichan + 1, result, port)
  end

  def decode_f32(anzf32, <<f::32-float, rest::binary>>, ichan, result, port) when anzf32 > 0 do
    val = Float.round(f, 8)
    unit = ltx_units(port, length(result))
    cv =  %{channel: ichan, value: val, unit: unit, prec: "F32"}

    result = result ++ [cv]
    decode_f32(anzf32 - 1, rest, ichan + 1, result, port)
  end

  # n * F16 entnehmen
  def decode_f16(0, rest, ichan, result, port) do
    decode_channels(rest, ichan, result, port)
  end

  def decode_f16(anzf16, <<0x3F::size(6), errno::size(10), rest::binary>>, ichan, result, port)
      when anzf16 > 0 do
    errmsg = ltx_error(errno)
    unit = ltx_units(port, length(result))
    cv =  %{channel: ichan, msg: errmsg, unit: unit, prec: "F16"}

    result = result ++ [cv]
    decode_f16(anzf16 - 1, rest, ichan + 1, result, port)
  end

  def decode_f16(anzf16, <<u16::size(16), rest::binary>>, ichan, result, port) when anzf16 > 0 do
    val = decode_float16(u16)
    unit = ltx_units(port, length(result))
    cv =   %{channel: ichan, value: val, unit: unit, prec: "F16"}

    result = result ++ [cv]
    decode_f16(anzf16 - 1, rest, ichan + 1, result, port)
  end

  # Bitweise HK entnehmen, skip HK wenn Bit 0
  def decode_hk(_, 0, rest, ichan, result, port) do
    decode_channels(rest, ichan, result, port)
  end

  def decode_hk(hk16bits, hcnt, rest, ichan, result, port)
      when hcnt > 0 and (hk16bits &&& 1) == 0 do
    decode_hk(hk16bits >>> 1, hcnt - 1, rest, ichan + 1, result, port)
  end

  def decode_hk(hk16bits, hcnt, <<u16::size(16), rest::binary>>, ichan, result, port)
      when hcnt > 0 do
    val = decode_float16(u16)
    unit = hk_unit(ichan)
    cv = %{channel: ichan, value: val, unit: unit, prec: "F16"}
    result = result ++ [cv]
    decode_hk(hk16bits >>> 1, hcnt - 1, rest, ichan + 1, result, port)
  end

  # --Hilfsfunktionen--
  # Fehlermeldungen im Klartext in FLoat16/32
  def ltx_error(errno) do
    case errno do
      1 -> "NoValue"
      2 -> "NoReply"
      3 -> "OldValue"
      6 -> "ErrorCRC"
      7 -> "DataError"
      8 -> "NoCachedValue"
      _ -> "Err#{errno}"
    end
  end

  # Einheiten fuer Port und Index (repeating)
  def ltx_units(port, idx) do
    units =
      case port do
        10 -> ["°C"]
        11 -> ["%rH", "°C"]
        12 -> ["Bar", "°C"]
        13 -> ["m", "°C"]
        14 -> ["m", "dBm"]
        15 -> ["°C", "uS/cm"]
        _ -> ["(nn)"]  # not known
      end

    anz = length(units)
    uidx = if anz > 1, do: rem(idx, anz), else: 0
    Enum.at(units, uidx)
  end

  # Einheit / Beschreibung HK-Kanaele
  def hk_unit(channel) do
    case channel do
      90 -> "V (HK_Bat)"
      91 -> "°C (HK_intTemp)"
      92 -> "%rH (HK_intHum.)"
      93 -> "mAh (HK_usedEnergy)"
      94 -> "mBar (HK_Baro)"
      _ -> "HK_unknown"
    end
  end

  # Dekodierung eines 16-Bit-Werts in einen Float
  def decode_float16(raw) do
    sign = if((raw &&& 0x8000) != 0, do: -1, else: 1)
    exponent = raw >>> 10 &&& 0x1F
    fraction = raw &&& 0x03FF

    cond do
      exponent == 0 ->
        Float.round(sign * :math.pow(2, -14) * (fraction / :math.pow(2, 10)), 6)

      exponent == 0x1F ->
        :infinity

      true ->
        Float.round(sign * :math.pow(2, exponent - 15) * (1 + fraction / :math.pow(2, 10)), 6)
    end
  end

  ## ----------------- TEST-CONSOLE START (entfernen fuer ELEMENTS) -----------------
  def test_decoder(hexstring, meta) do
    hexstring
    |> String.replace(~r/\s+/, "")
    |> String.codepoints()
    |> Enum.chunk_every(2)
    |> Enum.map(fn
      [a, b] -> String.to_integer(a <> b, 16)
      _ -> 0
    end)
    |> :binary.list_to_bin()
    |> parse(meta)
  end

  # Hauptfunktion für Tests - Keine IO.puts() im Parser, sonst wirf ELEMENTS einen raus
  def main do
    [
      # "13 01 41A4E148  9F 42A3 4DA8 5172 2B3F 63E8",
      #"130141A4E1489F42A34DA851722B3F63E8",
      #"12 42 4D24 4CE4 01 41948F5C 88 355B",
      "72 00 11 42 FC02 FC02 01FF800002884479"
    ]
    |> Enum.each(fn msg ->
      # Testport
      testport = 1
      IO.puts("Test Payload: #{msg} Port: #{testport}")

      msg
      |> test_decoder(%{meta: %{frame_port: testport}})
      |> IO.inspect(label: "Decoded Result")
    end)
  end

  ## ----------------- TEST-CONSOLE ENDE (entfernen fuer ELEMENTS) -----------------
end

## --- (entfernen fuer ELEMENTS):
Parser.main()
